ddsrouter_core/src/cpp/communication/rpc/RPCBridge.cpp

```diff
@@ -0,0 +1,352 @@
+            {
+                // info or debuglogInfo(DDSROUTER_RPCBRIDGE, "No data found at service Reader in topic " << reader->topic()
+                                                                        // << ". Finish transmission.");
+
+                tasks_map_[reader->guid()].first = false;
+                return;
+            }
+        }
+
+        if (!ret)
+        {
+            // Error reading data
+            logWarning(DDSROUTER_RPCBRIDGE, "Error taking data at service Reader in topic " << reader->topic()
+                                                                        << ". Error code " << ret
+                                                                      << ". Skipping data and continue.");
+            continue;
+        }
+
+        // logDebug(DDSROUTER_RPCBRIDGE,
+        //         "Track " << reader_participant_id_ << " for topic " << topic_ <<
+        //         " transmitting data from remote endpoint " << data->source_guid << ".");
+
+        // different logs for request and reply
+
+        if (RPCTopic::is_request_topic(reader->topic()))
+        {
+            for (auto& service_registry : service_registries_)
+            {
+                // TODO: Include is_repeater() condition
+                if (data->receiver_participant_id == service_registry.first || !service_registry.second->enabled())
+                {
+                    continue;
+                }
+
+                ret = request_writers_[service_registry.first]->write(data);
+
+                if (!ret)
+                {
+                    // logWarning(DDSROUTER_RPCBRIDGE, "Error writting data in Track " << topic_ << ". Error code "
+                    //                                                             << ret <<
+                    //         ". Skipping data for this writer and continue.");
+                    continue;
+                }
+            }
+        }
+        else if (RPCTopic::is_reply_topic(reader->topic()))
+        {
+            std::pair<ParticipantId, SampleIdentity> registry_entry = service_registries_[reader->participant_id()]->get(data->write_params.sample_identity().sequence_number());
+
+            if (!registry_entry.first.is_valid())
+            {
+                // Request already replied by another server connected to the same participant as this one
+                payload_pool_->release_payload(data->payload);
+                continue;
+            }
+
+            ret = reply_writers_[registry_entry.first]->write(data, registry_entry.second);
+
+            if (!ret)
+            {
+                // logWarning(DDSROUTER_RPCBRIDGE, "Error writting data in Track " << topic_ << ". Error code "
+                //                                                             << ret <<
+                //         ". Skipping data for this writer and continue.");
+                payload_pool_->release_payload(data->payload);
```

Comment:
I would try to join every release_payload in one call (last one) and intermediate "error cases" all flow to it.
This is a common source of errors where changing/removing this `release`clause will create very hard trackable errors